src/include/stir/listmode/CListRecord.h

```diff
@@ -107,107 +96,34 @@ class CListEvent
 
-
-//! A class for storing and using a timing record from a listmode file
-/*! \ingroup listmode
-    CListTime is used to provide an interface to the 'timing' events 
-    in the list mode stream. Usually, the timing event also contains 
-    gating information. For rotating scanners, it could also contain
-    angle info.
-
-    \todo this is still under development. Things to add are angles
-    or so for rotating scanners. Also, some info on the maximum
-    (and actual?) number of gates would be useful.
-    \see CListModeData for more info on list mode data. 
-*/
-class CListTime
+class CListTime : public ListTime
 {
 public:
   virtual ~CListTime() {}
-
-  virtual unsigned long get_time_in_millisecs() const = 0;
-  inline double get_time_in_secs() const
-    { return get_time_in_millisecs()/1000.; }
-
-  virtual Succeeded set_time_in_millisecs(const unsigned long time_in_millisecs) = 0;
-  inline Succeeded set_time_in_secs(const double time_in_secs)
-    { 
-      unsigned long time_in_millisecs;
-      round_to(time_in_millisecs, time_in_secs/1000.);
-      return set_time_in_millisecs(time_in_millisecs); 
-    }
-
 };
 
-//! A class recording external input to the scanner (normally used for gating)
-/*! For some scanners, the state of some external measurements can be recorded in the
-   list file, such as ECG triggers etc. We currently assume that these take discrete values.
-
-   If your scanner has more data available, you can provide it in the derived class.
-*/
-class CListGatingInput
+class CListGatingInput : public ListGatingInput
 {
 public:
   virtual ~CListGatingInput() {}
-
-  //! get gating-related info
-  /*! Generally, gates are numbered from 0 to some maximum value.
-   */
-  virtual unsigned int get_gating() const = 0;
-
-  virtual Succeeded set_gating(unsigned int) = 0;
 };
 
-//! A class for a general element of a list mode file
-/*! \ingroup listmode
-    This represents either a timing or coincidence event in a list mode
-    data stream.
-
-    Some scanners can have more types of records. For example,
-    the Quad-HiDAC puts singles information in the
-    list mode file. If you need that information,
-    you will have to do casting to e.g. CListRecordQHiDAC.
-    
-    \see CListModeData for more info on list mode data. 
-*/
-class CListRecord
+class CListRecord : public ListRecord
 {
 public:
   virtual ~CListRecord() {}
-
-  virtual bool is_time() const = 0;
-
-  virtual bool is_event() const = 0;
-
-  virtual CListEvent&  event() = 0;
-  virtual const CListEvent&  event() const = 0;
-  virtual CListTime&   time() = 0; 
-  virtual const CListTime&   time() const = 0; 
-
-  virtual bool operator==(const CListRecord& e2) const = 0;
-  bool operator!=(const CListRecord& e2) const { return !(*this == e2); }
+//  virtual bool operator==(const CListRecord& e2) const =0;
+//  bool operator!=(const CListRecord& e2) const { return !(*this == e2); }
 
 };
 
 class CListRecordWithGatingInput : public CListRecord
```

Comment:
This seems tricky. It needs to be derived from `CListRecord` but really also from `ListRecordWithGatingInput`. It seems therefore that we need to use virtual inheritance (i.e. `CListRecord: public virtual ListRecord` for every class that derived from `ListRecord`). hmmm...

Reply:
sorry. hard to understand this for me. I don't know what the state of the code was, and even not where this occured as the line number doesn't mean anything to me. Maybe you can create a permalink and insert the line, and say where you tried to put the `virtual`